firewood/src/merkle/parallel.rs

```diff
@@ -0,0 +1,500 @@
+#[derive(Debug)]
+pub struct ParallelMerkle {
+    workers: [Option<WorkerSender>; BranchNode::MAX_CHILDREN],
+}
+
+impl Default for ParallelMerkle {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ParallelMerkle {
+    /// Default constructor
+    #[must_use]
+    pub fn new() -> Self {
+        ParallelMerkle {
+            workers: [(); BranchNode::MAX_CHILDREN].map(|()| None),
+        }
+    }
+
+    /// Normalize the root to allow clean separation of the trie into an array of subtries that
+    /// can be operated on independently by the worker threads.
+    fn normalize_root(
+        &self,
+        proposal: &mut NodeStore<MutableProposal, FileBacked>,
+    ) -> Box<BranchNode> {
+        // There are 3 different cases to handle depending on the value of the root node.
+        //
+        // 1. If root is None, create a branch node with an empty partial path and a None for
+        //    value. Create Nones for all of its children.
+        // 2. If the existing root has a partial path, then create a new root with an empty
+        //    partial path and a None for a value. Push down the previous root as a child. Note
+        //    that this modified Merkle trie is not currently valid and may need to be updated
+        //    during the post-processing step.
+        // 3. If the existing root does not have a partial path, then there is nothing we need
+        //    to do if it is a branch. If it is a leaf, then convert it into a branch.
+        //
+        // Cases 2 and 3 are handled by `normalize_for_insert`. This function returns a branch
+        // node with an empty partial path.
+        let root_node = proposal.root_mut().take();
+        if let Some(node) = root_node {
+            return node.normalize_for_insert();
+        }
+        // Empty trie. Create a branch node with an empty partial path and a None for a value.
+        BranchNode {
+            partial_path: Path::new(),
+            value: None,
+            children: BranchNode::empty_children(),
+        }
+        .into()
+    }
+
+    /// After performing parallel modifications, it may be necessary to perform post processing
+    /// to return the Merkle trie to the correct canonical form.
+    fn postprocess_trie(
+        &self,
+        nodestore: &mut NodeStore<MutableProposal, FileBacked>,
+        mut branch: Box<BranchNode>,
+    ) -> Result<Option<Node>, FileIoError> {
+        // Check if the Merkle trie has an extra root node. If it does, apply transform to
+        // return trie to a valid state by following the steps below:
+        //
+        // If the root node has:
+        // 0 children and no value, the trie is empty. Just delete the root.
+        // 0 children and a value (from an empty key), the root should be a leaf
+        // 1 child and no value, the child should be the root (need to update partial path)
+        // In all other cases, the root is already correct.
```

Comment:
Move this to the doc block around line 120 instead of just a regular comment. If we do generate docs for this, it would be nice to have the explanation in the doc comment.